src/opf/ed.jl

```diff
@@ -0,0 +1,467 @@
+    max_ptdf_iterations > 0 || error("EconomicDispatch option max_ptdf_iterations must be a positive integer")
+    max_ptdf_per_iteration > 0 || error("EconomicDispatch option max_ptdf_per_iteration must be a positive integer")
+
+    # Grab some data
+    N = length(ref[:bus])
+    G = length(ref[:gen])
+    E = length(data["branch"])
+    L = length(ref[:load])
+
+    pmin = [ref[:gen][g]["pmin"] for g in 1:G]
+    pmax = [ref[:gen][g]["pmax"] for g in 1:G]
+    PD = sum(ref[:load][l]["pd"] for l in 1:L)
+    pfmax = [data["branch"]["$e"]["rate_a"] for e in 1:E]
+
+    minimum_reserve = get(data, "minimum_reserve", 0.0)
+    rmin = (minimum_reserve > 0.0) ? [ref[:gen][g]["rmin"] for g in 1:G] : zeros(T, G)
+    rmax = (minimum_reserve > 0.0) ? [ref[:gen][g]["rmax"] for g in 1:G] : zeros(T, G)
+
+    model = JuMP.GenericModel{T}(optimizer)
+    model.ext[:opf_model] = EconomicDispatch  # for internal checks
+    model.ext[:solve_metadata] = Dict{Symbol,Any}(
+        :iterative_ptdf => iterative_ptdf,
+        :iterative_ptdf_tol => iterative_ptdf_tol,
+        :max_ptdf_iterations => max_ptdf_iterations,
+        :max_ptdf_per_iteration => max_ptdf_per_iteration,
+    )
+
+    # Variables
+
+    JuMP.@variable(model, pmin[g] <= pg[g in 1:G] <= pmax[g])
+    JuMP.@variable(model, rmin[g] <= r[g in 1:G] <= rmax[g])
+    JuMP.@variable(model, pf[e in 1:E])
+
+    # for upper bounds, see https://github.com/AI4OPT/OPFGenerator/pull/64/files#r1529320645
+    JuMP.@variable(model, δpb_surplus >= 0)
+    JuMP.@variable(model, δpb_shortage >= 0)
+    JuMP.@variable(model, δr_shortage >= 0)
+    JuMP.@variable(model, δf[1:E] >= 0)
+
+    JuMP.@constraint(model,
+        pf_lower_bound[e in 1:E],
+        pf[e] + δf[e] >= -pfmax[e] 
+    )
+    JuMP.@constraint(model,
+        pf_upper_bound[e in 1:E],
+        pf[e] - δf[e] <= pfmax[e] 
+    )
+
+    if !soft_power_balance
+        JuMP.set_upper_bound(δpb_surplus, 0)
+        JuMP.set_upper_bound(δpb_shortage, 0)
+    end
+
+    if !soft_reserve_requirement && minimum_reserve > 0.0
+        JuMP.set_upper_bound(δr_shortage, 0)
+    end
+
+    if !soft_thermal_limit
+        JuMP.set_upper_bound.(δf, 0)
+    end
+
+    # Constraints
+
+    JuMP.@constraint(model,
+        total_generation[g in 1:G],
+        pg[g] + r[g] <= pmax[g]
+    )
+
+    JuMP.@constraint(model,
+        power_balance,
+        sum(pg) + δpb_surplus - δpb_shortage == PD
+    )
+
+    JuMP.@constraint(model,
+        reserve_requirement,
+        sum(r) + δr_shortage >= minimum_reserve
+    )
+
+    model.ext[:PTDF] = PM.calc_basic_ptdf_matrix(data)
+    model[:ptdf_flow] = Vector{JuMP.ConstraintRef}(undef, E)
+    model.ext[:tracked_branches] = zeros(Bool, E)
+    model.ext[:ptdf_iterations] = 0
+
+    # Objective
+
+    c = [ref[:gen][g]["cost"][2] for g in 1:G]
+    q = [get!(ref[:gen][g], "reserve_cost", zeros(T, 3))[2] for g in 1:G]
```

Comment:
Does your sampler also generate reserve costs?

Reply:
Why the in-place `get!` as opposed to non-modifying `get`?

Reply:
For the test a few lines down. I'll move this to the E2ELR sampler (it will fill in zeros for reserve cost)

Reply:
Note I am still supporting no reserve data by assuming there should be no reserves. This allows to build the model without going through an OPFSampler.

Perhaps I should add some info/warning or even error if no reserve data is found?

Reply:
I think you're making your life harder than it needs to be :)

PGLib cases do not have reserves. Currently we only have the E2ELR reserve sampler, which also doesn't assign any cost to reserves. So you're supporting a quadratic reserve cost that, in the current state of things, would never occur in practice. That was the motivation for the question "does your sampler also generate reserve costs?"

Reply:
> I am still supporting no reserve data by assuming there should be no reserves if no data is found

I think that's good. Fool-proof :)

The `get` vs `get!` part relates to whether we modify in the input data or not. I don't know whether PowerModels' `build_ref` creates a copy of the data dictionary. If so, we're good. If not, we're silently modifying the input data, and I don't know how I feel about that.